src/printer.js

```diff
@@ -3859,6 +3859,10 @@ function makeString(rawContent, enclosingQuote) {
   // Matches _any_ escape and unescaped quotes (both single and double).
   const regex = /\\([\s\S])|(['"])/g;
 
+  // Matches any unnecessarily escaped character.
+  // Adapted from https://github.com/eslint/eslint/blob/de0b4ad7bd820ade41b1f606008bea68683dc11a/lib/rules/no-useless-escape.js#L27
+  const regexUnnecessaryStringEscapes = /(^|[^\\])\\([^\\nrvtbfux\r\n\u2028\u2029"'0])/;
```

Comment:
This is missing some cases from [Annex B](https://tc39.github.io/ecma262/#prod-annexB-LegacyOctalEscapeSequence), which extends the definition of `EscapeSequence` to include octals in sloppy mode. So the latter character class needs to end with `0-7`, not `0`. (In strict mode, octal escapes other than `\0` are a syntax error, so no further accounting for them is necessary.)

Otherwise I believe this captures all of the characters for which putting a backslash before them does something other than just giving you the character.

However, this still isn't right, since it won't match, e.g., `"\\\a"`: the initial "not preceded by a backslash" logic prevents it. That won't cause bugs, but it will cause some unnecessary escapes to be preserved.